pyorbital/orbital.py

```diff
@@ -160,11 +160,145 @@ class Orbital(object):
+
+            if iindex in (0, len(dates)):
+                if iindex == len(dates):
+                    # Reset index if beyond the right boundary (see note 2. above)
+                    iindex -= 1
+            elif abs(sdate - dates[iindex - 1]) < abs(sdate - dates[iindex]):
+                # Choose the closest of the two surrounding dates
+                iindex -= 1
+
+            # Make sure the TLE we found is within the threshold
+            delta_days = abs(sdate - dates[iindex]) / np.timedelta64(1, 'D')
+            tle_thresh = 7
+            if delta_days > tle_thresh:
+                raise IndexError(
+                    "Can't find tle data for %s within +/- %d days around %s" %
+                    (self.satellite_name, tle_thresh, sdate))
+
+            if delta_days > 3:
+                warnings.warn("Found TLE data for %s that is %f days apart" %
+                              (sdate, int(delta_days)))
+
+            # Select TLE data
+            tle1 = tle_data[iindex * 2]
+            tle2 = tle_data[iindex * 2 + 1]
+            self._tle = tlefile.read(self.satellite_name, tle_file=None,
+                                     line1=tle1, line2=tle2)
+
+        # Not using TLE archive;
+        # Either using current celestrek TLE,
+        # or using user supplied Line 1 and line 2
+        else:
+            sat_time = self.tle2datetime64(float(self._tle.line1[18:32]))
+            # Object just created
+            if not self.utctime:
+                self.utctime = datetime.now()
+            mismatch = sat_time.astype(datetime) - self.utctime
+            if abs(mismatch.days) > 30:
+                raise IndexError("""
+                    Current TLE from celestrek is %d days newer than
+                    requested orbital parameter. Please use TLE archive
+                    for accurate results""" % (mismatch.days))
+            if abs(mismatch.days) > 7:
+                warnings.warn("""
+                    Current TLE from celestrek is %d days newer than
+                    requested orbital parameter. Please use TLE archive
+                    for accurate results""" % (mismatch.days))
+
+        return self._tle
+
+    @tle.setter
+    def tle(self, new_tle):
+        self._tle = new_tle
+
+    @property
+    def orbit_elements(self):
+        return OrbitElements(self.tle)
+
+    @orbit_elements.setter
+    def orbit_elements(self, new_orbit_elements):
+        self._orbit_elements = new_orbit_elements
+
+    @property
+    def sgdp4(self):
+        return _SGDP4(self.orbit_elements)
+
+    @sgdp4.setter
+    def sgdp4(self, new_sgdp4):
         self._sgdp4 = _SGDP4(self.orbit_elements)
 
+    @property
+    def utctime(self):
+        return self._utctime
+
+    @utctime.setter
+    def utctime(self, utc_time):
+        if not isinstance(utc_time, datetime):
+            times = np.array(utc_time, dtype='datetime64[m]')
+            if times.max() - times.min() > np.timedelta64(3, 'D'):
```

Comment:
Here, could we check instead that both min and max times are within 3 days of the epoch instead? or do you think it would be too restrictive?